Index: pjproject/_source/pjlib-util/src/pjlib-util/resolver.c
===================================================================
--- pjproject.orig/_source/pjlib-util/src/pjlib-util/resolver.c
+++ pjproject/_source/pjlib-util/src/pjlib-util/resolver.c
@@ -1384,9 +1384,14 @@ static void report_nameserver_status(pj_
 	q_id = (pj_uint32_t)-1;
     }
 
-    if (!pkt || rcode == PJ_DNS_RCODE_SERVFAIL ||
-	        rcode == PJ_DNS_RCODE_REFUSED ||
-	        rcode == PJ_DNS_RCODE_NOTAUTH) 
+    /* Some nameserver is reported to respond with PJ_DNS_RCODE_SERVFAIL for
+     * missing AAAA record, and the standard doesn't seem to specify that
+     * SERVFAIL should prevent the server to be contacted again for other
+     * queries. So let's not mark nameserver as bad for SERVFAIL response.
+     */
+    if (!pkt || /* rcode == PJ_DNS_RCODE_SERVFAIL || */
+            rcode == PJ_DNS_RCODE_REFUSED ||
+            rcode == PJ_DNS_RCODE_NOTAUTH)
     {
 	is_good = PJ_FALSE;
     } else {
